### PR TITLE
(WIP) EOS-26312:enable di flag s3 api motr side changes

### PR DESCRIPTION
**Pull request by rajatpatil98**
# Problem Statement
- Make Motr side changes to Enable DI in main using the s3 flag

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: ioservice/io_foms.c [hunk 1]
```diff
@@ -2077,16 +2077,19 @@ static int stob_io_create(struct m0_fom *fom)
...
-			m0_free(fom_obj->fcrw_stio);
-			return M0_ERR(-ENOMEM);
+			if (m0_buf_alloc( &rw_replyfop->rwr_di_data_cksum,
+					  rw_replyfop->rwr_di_data_cksum.b_nob) != 0) {
+				m0_free(fom_obj->fcrw_stio);
+				return M0_ERR(-ENOMEM);
+			}
+		} else {
+			rw_replyfop->rwr_di_data_cksum.b_nob  = 0;
+			rw_replyfop->rwr_di_data_cksum.b_addr = NULL;
```
Comment (madhavemuri) [inline ioservice/io_foms.c:2090]: This part can be moved to fom_create.

File: ioservice/io_foms.c [hunk 2]
```diff
@@ -2077,16 +2077,19 @@ static int stob_io_create(struct m0_fom *fom)
...
 		}
-	}
-	else {
+	} else {
 		rw_replyfop->rwr_di_data_cksum.b_nob  = 0;
 		rw_replyfop->rwr_di_data_cksum.b_addr = NULL;
```
Comment (madhavemuri) [inline ioservice/io_foms.c:2094]: Same for this block as well.

File: lib/cksum.c [hunk 3]
```diff
@@ -238,6 +237,22 @@ bool m0_calc_verify_cksum_one_unit(struct m0_generic_pi *pi,
 
 M0_EXPORTED(m0_calc_verify_cksum_one_unit);
 
+bool m0_is_di_enabled(struct m0_op_io *ioo) {
+	return (ioo->ioo_obj->ob_entity.en_flags & M0_ENF_DI);
```
Comment (madhavemuri) [inline lib/cksum.c:241]: Brackets are not needed.

File: lib/cksum.c [hunk 4]
```diff
@@ -238,6 +237,22 @@ bool m0_calc_verify_cksum_one_unit(struct m0_generic_pi *pi,
 
 M0_EXPORTED(m0_calc_verify_cksum_one_unit);
 
+bool m0_is_di_enabled(struct m0_op_io *ioo) {
```
Comment (madhavemuri) [inline lib/cksum.c:240]: Move open curly braces to next line

File: lib/cksum.c [hunk 5]
```diff
@@ -238,6 +237,22 @@ bool m0_calc_verify_cksum_one_unit(struct m0_generic_pi *pi,
...
+	return (ioo->ioo_obj->ob_entity.en_flags & M0_ENF_DI);
+}
+
+M0_EXPORTED(m0_is_di_enabled);
+
+bool m0_is_checksum_validation_allowed(struct m0_op_io *ioo)
+{
+	if (m0_is_di_enabled(ioo) && !ioo->ioo_dgmode_io_sent)
+		return true;
+	else
```
Comment (madhavemuri) [inline lib/cksum.c:250]: `else` can be removed.

File: lib/cksum_utils.c [hunk 6]
```diff
@@ -139,14 +139,14 @@ M0_INTERNAL void * m0_extent_vec_get_checksum_addr(void *cksum_buf_vec,
...
 
 	/** Assert to make sure the the offset is lying within the extent */
-	M0_ASSERT(i < vec->iv_vec.v_nr);
-
-	/** get the checksum_addr */
-	m0_bufvec_cursor_init(&cksum_cursor, cksum_vec);
-
-	if (attr_nob) {
-		m0_bufvec_cursor_move(&cksum_cursor, attr_nob);
+	// M0_ASSERT(i < vec->iv_vec.v_nr);
```
Comment (madhavemuri) [inline lib/cksum_utils.c:142]: Remove this statement or change the comment to c-style.

File: lib/cksum_utils.c [hunk 7]
```diff
@@ -139,14 +139,14 @@ M0_INTERNAL void * m0_extent_vec_get_checksum_addr(void *cksum_buf_vec,
...
+	if (i < vec->iv_vec.v_nr) {
+		/** get the checksum_addr */
+		m0_bufvec_cursor_init(&cksum_cursor, cksum_vec);
+		if (attr_nob != 0) {
```
Comment (madhavemuri) [inline lib/cksum_utils.c:146]: is this check needed ?
cursor_move can work with 0 as well.

File: motr/client.h [hunk 8]
```diff
@@ -630,7 +630,8 @@ enum m0_entity_type {
 	 * If this flags is set during entity_create() that means application
 	 * do not support update operation. This flag is not in use yet.
 	 */
-	M0_ENF_NO_RMW =  1 << 1
+	M0_ENF_NO_RMW =  1 << 1,
+	M0_ENF_DI = 1 << 2
```
Comment (madhavemuri) [inline motr/client.h:634]: Add a description for this flag.

File: motr/io_nw_xfer.c [hunk 9]
```diff
@@ -647,41 +650,41 @@ static void target_ioreq_seg_add(struct target_ioreq              *ti,
 		                      + ioo->ioo_ext.iv_vec.v_count[ioo->ioo_ext.iv_vec.v_nr - 1];
 		/* If ioo_attr struct is not allocated then skip checksum computation */
 		is_goff_in_range = m0_ext_is_in(&goff_span_ext, goff) &&
-		                                ioo->ioo_attr.ov_vec.v_nr;
+		                        (ioo->ioo_obj->ob_entity.en_flags & M0_ENF_DI);
```
Comment (madhavemuri) [inline motr/io_nw_xfer.c:653]: Use m0_is_di_enabled() here and other places.

File: ioservice/io_foms.c [hunk 10]
```diff
@@ -2077,16 +2077,19 @@ static int stob_io_create(struct m0_fom *fom)
 						           rwfop->crw_cksum_size);
 		}
 
-		// Its expected to receive atleast on unit start in a fop
-		M0_ASSERT(rw_replyfop->rwr_di_data_cksum.b_nob > 0);
+		/* Its expected to receive atleast on unit start in a fop */
```
Comment (madhavemuri) [inline ioservice/io_foms.c:2080]: Can you re-check this command ?

File: motr/io_req_fop.c [hunk 11]
```diff
@@ -178,6 +179,15 @@ static void application_attribute_copy(struct m0_indexvec *rep_ivec,
 			return;
 	}
 	off = ti_cob_index % unit_size;
+	if (off != 0) {
+		if (!m0_ivec_cursor_move(&ti_cob_cursor, unit_size - off)) {
+			ti_cob_index = m0_ivec_cursor_index(&ti_cob_cursor);
+		} else {
+			/** invalid cusror position */
```
Comment (madhavemuri) [inline motr/io_req_fop.c:186]: Add M0_IMPOSSIBLE here.

File: motr/io_req_fop.c [hunk 12]
```diff
@@ -208,11 +218,12 @@ static void application_attribute_copy(struct m0_indexvec *rep_ivec,
 		while (ti_cob_index != rep_index) {
 			if (m0_ivec_cursor_move(&ti_cob_cursor, unit_size) ||
 			    m0_ivec_cursor_move(&ti_goff_cursor, unit_size)) {
-				M0_ASSERT(0);
-			}
+					M0_ASSERT(0);
```
Comment (madhavemuri) [inline motr/io_req_fop.c:221]: Instead of assert, add M0_IMPOSSIBLE here.

**Comment by rajatpatil98** [comment]
Closing this PR, This patch is covered in https://github.com/Seagate/cortx-motr/pull/1345 and merged in main.

**Pull request closed without merging**